testsuite/testsuite.c

```diff
@@ -166,6 +166,55 @@ int test_spawn_prog(const char *prog, const char *const args[])
 	return EXIT_FAILURE;
 }
 
+int test_spawn_and_wait4_prog(const char *prog, const char *const args[])
+{
+	int err, pid, ret;
+	const char *msg;
+
+	pid = fork();
+	if (pid < 0) { \
+		ERR("could not fork(): %m\n");
+		return 1;
+	}
+
+	if (pid == 0)
+		test_spawn_prog(prog, args);
+
+	do {
+		pid = wait(&err);
+		if (pid == -1) {
+			ERR("error waitpid(): %m\n");
+			return 1;
+		}
+	} while (!WIFEXITED(err) && !WIFSIGNALED(err));
+
+	if (WIFEXITED(err)) {
+		ret = WEXITSTATUS(err);
+		if (ret) {
+			msg = "exited with";
+			goto print_error;
+		}
+		return WEXITSTATUS(err);
+	} else if (WIFSIGNALED(err)) {
+		ret = WTERMSIG(err);
+		if (ret) {
+			msg = "signalled with";
+			goto print_error;
+		}
+		return WTERMSIG(err);
+	}
+
+	return 0;
+
+print_error:
+	ERR("%s %s %d\n", prog, msg, ret);
+	while (*args) {
+		LOG("  args: %s\n", *args);
```

Comment:
Nit: let's make this a small helper function. I think we want ERR for both program and args.